plugins/audit_logs/scripts/report.py

```diff
@@ -0,0 +1,87 @@
+"""
+This script is for creating download reports for a given folder.
+
+    girder shell --plugins=audit_logs report.py -- --folder=57557fac8d777f68be8f3f49 --start-date=2018-09-10T13:55:34.847Z --end-date=2018-09-13T13:55:34.847Z report.csv
+"""
+import click
+import datetime
+import dateutil.parser
+
+from bson.objectid import ObjectId
+
+from girder.models.item import Item
+from girder.models.folder import Folder
+from girder.constants import AccessType
+from girder.plugins.audit_logs import Record
+
+
+def index_folder(folderId):
+    if Folder().load(folderId, force=True) is None:
+        raise ValueError('folderId={} was not a valid folder'.format(folderId))
+    items = Item().find({'folderId': ObjectId(folderId)})
+    subfolders = Folder().find({'parentId': ObjectId(folderId)})
+
+    files = []
+    for item in items:
+        for file in Item().childFiles(item, fields={'_id': True}):
+            fileId = file['_id']
+            files.append(fileId)
+
+    for folder in subfolders:
+        files += index_folder(folder['_id'])
+
+    return files
+
+
+def get_file_download_records(files, start=None, end=None):
+    query = {
+        'type': 'file.download',
+        'details.fileId': {
+            '$in': files,
+        },
+        'details.startByte': 0,
+        'when': {
+            '$gte': start,
+            '$lt': end,
+        }
+    }
+    if (start is not None) or (end is not None):
+        whenClause = { 'when': {} }
+        if start is not None:
+            whenClause['when']['$gte'] = dateutil.parser.parse(start)
+        if end is not None:
+            whenClause['when']['$lt'] = dateutil.parser.parse(end)
+        query.update(whenClause)
+    return Record().find(query)
+
+
+def csv_report_generator(files, records):
+    """
+    :param files: a list of objectId strings\e
+    :param records: a mongo cursor with all download records
+    :return: an output generator
+    """
+    yield 'file_id,ip,timestamp'
+    for r in records:
+        yield '{id},{address},{timestamp}'.format(
+            id=r['details']['fileId'],
+            address=r['ip'],
+            timestamp=str(r['when']),
+        )
+
+
+@click.command()
+@click.option('--folder', help='folder ID to use as root for all download reports.', required=True)
+@click.option('--start-date', help='UTC beginning date to look for downloads. format: YYYY-MM-DDThh:mm:ss.sssZ')
+@click.option('--end-date', help='UTC end date to look for downloads.')
+@click.argument('output')
+def report(folder, start_date, end_date, output):
+    files = index_folder(folder)
+    records = get_file_download_records(files, start=start_date, end=end_date)
+    gen = csv_report_generator(files, records)
+    with open(output, 'w') as out:
+        out.writelines((line + '\n' for line in gen))
```

Comment:
Despite this being a CSV of very simple data types, I would still opt for using the `csv` library out of the stdlib for this. If we should ever add more complex types, escaping would work correctly, also it correctly adds the CRLF line terminators by default.